.php-cs-fixer.dist.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+/* Pi-hole: A black hole for Internet advertisements
+*  (c) 2022 Pi-hole, LLC (https://pi-hole.net)
+*  Network-wide ad blocking via your own hardware.
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license.
+*/
+
+declare(strict_types=1);
+
+$finder = PhpCsFixer\Finder::create()
+    ->ignoreDotFiles(false)
+    ->ignoreVCSIgnored(true)
+    ->exclude('scripts/vendor')
+    ->in(__DIR__)
+;
+
+$config = new PhpCsFixer\Config();
+$config
+    ->setRules(array(
+        '@PSR12' => true,
+        '@PhpCsFixer' => true,
+        'array_syntax' => array('syntax' => 'long'),
+    ))
+    ->setFinder($finder)
+;
+
```

Comment:
I think we should avoid `@PhpCsFixer`. This introduce a lot of unneeded blank lines and force the use of `single quotes`.

Reply:
Are there any other drawbacks? I don't see many blank lines being added, only around the `require_once` and that doesn't seem like an issue to me. Using single quotes is a good idea as, to my understanding, they are handled as constants in PHP and, hence, even somewhat faster than double quoted strings which are expanded both for escape sequences and variable substitution - possibly more.

Reply:
I created a second branch (only PSR-12, no `PhpCsFixer`) and see the following differences:

### Extra spaces and no `()` for `require_once`:
`PhpCsFixer` + `PSR-12`
```php
require_once 'scripts/pi-hole/php/FTL.php';

require_once 'scripts/pi-hole/php/password.php';

require_once 'scripts/pi-hole/php/database.php';

require_once 'scripts/pi-hole/php/auth.php';
```
`PSR-12` alone
``` php
require_once("scripts/pi-hole/php/FTL.php");
require_once("scripts/pi-hole/php/password.php");
require_once("scripts/pi-hole/php/database.php");
require_once("scripts/pi-hole/php/auth.php");
```
---

### Single quotes instead of double quotes and `die()` is not replaced by `exit()`
`PhpCsFixer` + `PSR-12`
``` php
    if (isset($_GET['auth'])) {
        if ($_GET['auth'] !== $pwhash) {
            exit('Not authorized!');
```
`PSR-12` alone
``` php 
    if (isset($_GET["auth"])) {
       if ($_GET["auth"] !== $pwhash) {
            die("Not authorized!");
         }
```
---

### Spacing between operators is enforced for readiness
`PhpCsFixer` + `PSR-12`
```php
        pihole_execute('disable '.$disable.'s');
        file_put_contents('../custom_disable_timer', ($timestamp + $disable) * 1000);
```
`PSR-12` alone
``` php
        pihole_execute("disable ".$disable."s");
        file_put_contents("../custom_disable_timer", ($timestamp+$disable)*1000);
```
---

### Code simplification (an unnecessary `else` block is removed)
`PhpCsFixer` + `PSR-12`
``` php
// @@ -38,25 +38,24 @@ function SQLite3_connect_try($filename, $mode, $trytoreconnect)
        // sqlite3 throws an exception when it is unable to connect, try to reconnect after 3 seconds
        if ($trytoreconnect) {
            sleep(3);

            return SQLite3_connect_try($filename, $mode, false);
        }

        // If we should not try again (or are already trying again!), we return the exception string
        // so the user gets it on the dashboard
        return $filename.': '.$exception->getMessage();
    }
}
```
`PSR-12` alone
``` php
// @@ -38,25 +38,24 @@ function SQLite3_connect_try($filename, $mode, $trytoreconnect)
        // sqlite3 throws an exception when it is unable to connect, try to reconnect after 3 seconds
        if ($trytoreconnect) {
            sleep(3);
            return SQLite3_connect_try($filename, $mode, false);
        } else {
            // If we should not try again (or are already trying again!), we return the exception string
            // so the user gets it on the dashboard
            return $filename.": ".$exception->getMessage();
        }
    }
}
```

---

To be honest, from what I've seen in our codebase, I'm leaning towards the combination of both, due to using single quotes and adding in extra readability around (math) operators. Removing unnecessary `else` blocks is just the icing on the cake and wouldn't have been needed to convince me.

Reply:
Another example concerning the spaces around operators:
``` diff
-function add_to_table($db, $table, $domains, $comment=null, $wildcardstyle=false, $returnnum=false, $type=-1)
+function add_to_table($db, $table, $domains, $comment = null, $wildcardstyle = false, $returnnum = false, $type = -1)
```

and something I've found in addition (being explicit about variable substitution):
``` diff
-    if ($table === "domain_audit") {
-        $querystr = "INSERT OR IGNORE INTO $table ($field) VALUES (:$field);";
-    } elseif ($type === -1) {
-        $querystr = "INSERT OR IGNORE INTO $table ($field,comment) VALUES (:$field, :comment);";
+    if ('domain_audit' === $table) {
+        $querystr = "INSERT OR IGNORE INTO {$table} ({$field}) VALUES (:{$field});";
+    } elseif (-1 === $type) {
+        $querystr = "INSERT OR IGNORE INTO {$table} ({$field},comment) VALUES (:{$field}, :comment);";
```

I like the idea of putting the variable explicitly in `{}` just as we do in `bash` like `${abc}` to avoid issues with spilling. Here you can also see that double quotes are kept where they are useful.

(`-` is with `PSR-12`, `+` is with both)

Reply:
You are actually using `@PhpCsFixer`.

`@PhpCsFixer` and `@Symfony` are custom configs.
`@PhpCsFixer` = `@Symfony` config + extra config.
`@Symfony` = `@PSR12` + extra config.